nixos/modules/services/torrent/rtorrent.nix

```diff
@@ -0,0 +1,208 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.rtorrent;
+
+in {
+  options.services.rtorrent = {
+    enable = mkEnableOption "rtorrent";
+
+    dataDir = mkOption {
+      type = types.str;
+      default = "/var/lib/rtorrent";
+      description = ''
+        The directory where rtorrent stores its data files.
+      '';
+    };
+
+    downloadDir = mkOption {
+      type = types.str;
+      default = "${cfg.dataDir}/download";
+      description = ''
+        Where to put downloaded files.
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "rtorrent";
+      description = ''
+        User account under which rtorrent runs.
+      '';
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "rtorrent";
+      description = ''
+        Group under which rtorrent runs.
+      '';
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.rtorrent;
+      defaultText = "pkgs.rtorrent";
+      description = ''
+        The rtorrent package to use.
+      '';
+    };
+
+    port = mkOption {
+      type = types.int;
+      default = 50000;
+      description = ''
+        The rtorrent port.
+      '';
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to open the firewall for the port in <option>services.rtorrent.port</option>.
+      '';
+    };
+
+    rpcSocket = mkOption {
+      type = types.str;
+      default = "/run/rtorrent/rpc.sock";
+      description = ''
+        RPC socket path.
+      '';
+    };
+
+    extraConfig = mkOption {
```

Comment:
While not required I feel like a `settings` option as described by https://github.com/NixOS/rfcs/pull/42 would really bring this module to the next level.

Reply:
Yes, I'll do this.

Reply:
From taking a look the configuration file I agree, it does look quite complex! One thing you could do is add a `configFile` option, which can be used to provide the customization you are looking for.

What do you think @Infinisil?

Reply:
Looking at the example in https://rtorrent-docs.readthedocs.io/en/latest/cookbook.html#modernized-configuration-template, it pretty much looks to be an imperative programming language. I think an option `configText` of type `types.lines` would make sense, with the default configuration being defined as 
```
  config.services.rtorrent.configText = mkBefore ''
    # Instance layout (base paths)
    method.insert = cfg.basedir, private|const|string, (cat,"${cfg.dataDir}/")
    # ...
  '';
```

such that:
- If the user assigns to the option, it will come after the default config
- The user can override the whole config with a `mkForce` if necessary (so no hardcoded strings)

An additional `configFile` option could also be added, but other than users being able to set this to an impure path (which isn't really NixOS-y), I don't see much benefit in it, so I personally wouldn't do it if the `configText` idea works out.

Reply:
Perfect, feel free to suggest a clearer description for `configText`.